{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.rendered_html, p{direction:rtl; text-align:right !important}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(\"<style>.rendered_html, p{direction:rtl; text-align:right !important}</style>\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ابتدا کد زیر را برای مسئله‌ی بخش ۲.۱ زده شده. تابع aligner  یک تابع همترازی است که چیزی مابین همترازی سراسری و محلی است. در واقع در این الگوریتم هرچقد در رشته‌ی بزرگ‌تر (ژنوم) پیش برویم ولی در رشته‌ی کوچک‌تر (ژن) جلو نرویم هزینه‌ای نخواهد داشت. در این الگوریتم ارزش تطابق را ۱، عدم تطابق -۲ و فاصله را -۳ در نظر گرفتیم. سپس پس از پر کردن جدول برنامه‌نویسی پویای همترازی همانند همان الگوریتم تمرین درس، در آخر خانه‌ی از جدول که در سطر اخر قرار گرفته ( به معنی اینکه تا انتهای ژن جلو رفته‌ایم) را انتخاب میکنیم. همچنین با عقب رفتن از روی این الگوریتم تا اولین خانه‌ی که تطابق از آن جا شروع شده نقطه‌ی شروع را هم پیدا می‌کنیم و سپس به ازای دو به دوی ژن‌های ماربرگ و ژنوم گونه‌های مختلف ابولا این الگوریتم را صدا زده و نقطه‌ی شروع و پایان تطابق آن‌ها را در جدولی نگه میداریم.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#include<iostream>\n",
    "#include<fstream>\n",
    "\n",
    "using namespace std;\n",
    "\n",
    "int dp[9000][19000];\n",
    "//maximum size of genes and genome\n",
    "\n",
    "pair <int,int> aligner(string ref, string read){\n",
    "    int n = ref.size();\n",
    "    int m = read.size();\n",
    "    int match = 1;\n",
    "    int mismatch = -2;\n",
    "    int gap_open = -3;;\n",
    "    for (int i = 0; i < n+1; i++) {\n",
    "        dp[0][i] = 0;\n",
    "    }\n",
    "    for (int i = 1; i < m+1; i++) {\n",
    "        dp[i][0] = dp[i - 1][0] + gap_open;\n",
    "    }\n",
    "    //complete dp matrix\n",
    "    for (int i = 1; i < n+1; i++){\n",
    "        for (int j = 1; j < m+1; j++){\n",
    "            if (ref[i-1] == read[j-1]){\n",
    "                dp[j][i] = max(dp[j-1][i-1] + match, max(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "            if (ref[i-1] != read[j-1]){\n",
    "                dp[j][i] = max(dp[j-1][i-1] + mismatch, max(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    //find the index of cell with maximum point.\n",
    "    int maximum = -1e9;\n",
    "    int endIndexi = 0;\n",
    "    for (int i = 0; i < n +1; i++){\n",
    "        if (dp[m][i] > maximum) {\n",
    "            maximum = dp[m][i];\n",
    "            endIndexi = i;\n",
    "        }\n",
    "    }\n",
    "    //find the start point\n",
    "    int indexi = endIndexi;\n",
    "    int indexj = m;\n",
    "    while (indexj > 0) {\n",
    "\t\tif (dp[indexj][indexi] == dp[indexj - 1][indexi] + gap_open){indexj -= 1;}\n",
    "        else if(dp[indexj][indexi] == dp[indexj][indexi - 1] + gap_open){indexi -= 1;}\n",
    "        else if(dp[indexj][indexi] == dp[indexj - 1][indexi - 1] + mismatch && ref[indexi - 1] != read[indexj - 1]){\n",
    "            indexi -= 1;\n",
    "            indexj -= 1;\n",
    "        }\n",
    "        else if(dp[indexj][indexi] == dp[indexj - 1][indexi - 1] + match && ref[indexi - 1] == read[indexj - 1]){\n",
    "            indexi -= 1;\n",
    "            indexj -= 1;\n",
    "        }\n",
    "\n",
    "    }\n",
    "    int startIndexi = indexi;\n",
    "    return make_pair(startIndexi, endIndexi - 1);\n",
    "}\n",
    "int main() {\n",
    "    string genome[5];\n",
    "    string dict[] = {\"Reston_genome\", \"Sudan_genome\", \"TaiForest_genome\", \"Zaire_genome\", \"Bundibugyo_genome\"};\n",
    "    string dict2[] = {\"NP\", \"VP35\", \"VP40\", \"GP\", \"VP30\", \"VP24\", \"L\"};\n",
    "    for (int i = 0; i < 5; i++) {\n",
    "        ifstream cin(\"BioProjectFiles/\" + dict[i] + \".fasta\");\n",
    "        string s;\n",
    "        cin >> s;\n",
    "        cin >> s;\n",
    "        while (cin >> s) {\n",
    "            genome[i] += s;\n",
    "        }\n",
    "    }\n",
    "    ifstream cin(\"BioProjectFiles/Marburg_Genes.fasta\");\n",
    "    string genes[7];\n",
    "    string s;\n",
    "    cin >> s;\n",
    "    for (int i = 0; i < 7; i++) {\n",
    "        string gene = \"\";\n",
    "        while (cin >> s) {\n",
    "            if (s[0] == '>') {\n",
    "                break;\n",
    "            }\n",
    "            gene += s;\n",
    "        }\n",
    "        genes[i] = gene;\n",
    "    }\n",
    "    ofstream cout(\"a.txt\");\n",
    "    pair <int,int> alignes[5][7];\n",
    "    for (int i= 0; i < 7; i++){\n",
    "        for(int j = 0; j < 5; j++){\n",
    "            alignes[j][i] = aligner(genome[j], genes[i]);\n",
    "            cout << dict[j] << \" \" << dict2[i] << \" \" << alignes[j][i].first << \" \" << alignes[j][i].second <<endl;\n",
    "        }\n",
    "    }\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "نتایج کد بالا:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reston_genome NP 454 3170\n",
    "Sudan_genome NP 471 3101\n",
    "TaiForest_genome NP 490 3232\n",
    "Zaire_genome NP 449 3129\n",
    "Bundibugyo_genome NP 430 3237\n",
    "Reston_genome VP35 3085 4615\n",
    "Sudan_genome VP35 3046 4607\n",
    "TaiForest_genome VP35 3031 4627\n",
    "Zaire_genome VP35 3031 4615\n",
    "Bundibugyo_genome VP35 3007 4597\n",
    "Reston_genome VP40 4395 5756\n",
    "Sudan_genome VP40 4353 5731\n",
    "TaiForest_genome VP40 15372 16684\n",
    "Zaire_genome VP40 4389 5802\n",
    "Bundibugyo_genome VP40 9993 11359\n",
    "Reston_genome GP 5975 8750\n",
    "Sudan_genome GP 5939 8671\n",
    "TaiForest_genome GP 5965 8720\n",
    "Zaire_genome GP 5990 8680\n",
    "Bundibugyo_genome GP 5956 8683\n",
    "Reston_genome VP30 8373 9613\n",
    "Sudan_genome VP30 8332 9555\n",
    "TaiForest_genome VP30 8408 9625\n",
    "Zaire_genome VP30 8414 9625\n",
    "Bundibugyo_genome VP30 8385 9600\n",
    "Reston_genome VP24 10126 11338\n",
    "Sudan_genome VP24 10087 11349\n",
    "TaiForest_genome VP24 10159 11413\n",
    "Zaire_genome VP24 10166 11399\n",
    "Bundibugyo_genome VP24 10153 11374\n",
    "Reston_genome L 11356 18740\n",
    "Sudan_genome L 11333 18726\n",
    "TaiForest_genome L 11377 18758\n",
    "Zaire_genome L 11385 18785\n",
    "Bundibugyo_genome L 11387 18736\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "سپس برای بخش ۲.۳ یک کد تطابق سراسری زده شد که همانند بخش قبل است با این تفاوت که در صورت جلو رفتن در رشته‌ی ژنوم نیز باید هزینه‌ی گپ را قائل شویم، همچنین در این بخش باتوجه به اینکه فاصله‌ی ویرایش خواسته شده، امتیاز تطابق ۰، امتیاز عدم تطابق -۱ و امتیاز فاصله هم -۱ درنظر گرفته شده است. سپس دوباره به ازای هر دو ژنوم مختلف، بازه‌ی تطبیق داده شده با هر ژن را گرفته و دو به دو با یکدیگر همتراز میکنیم تا فاصله‌ی ویرایش پیدا شود. کد متناظر در ادامه آمده است.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#include<iostream>\n",
    "#include<fstream>\n",
    "\n",
    "using namespace std;\n",
    "\n",
    "int dp[19000][19000];\n",
    "\n",
    "int aligner(string ref, string read){\n",
    "    int n = ref.size();\n",
    "    int m = read.size();\n",
    "    int match = 0;\n",
    "    int mismatch = 1;\n",
    "    int gap_open = 1;\n",
    "    for (int i = 0; i < n+1; i++) {\n",
    "        dp[0][i] = i;\n",
    "    }\n",
    "    for (int i = 1; i < m+1; i++) {\n",
    "        dp[i][0] = i;\n",
    "    }\n",
    "    for (int i = 1; i < n+1; i++){\n",
    "        for (int j = 1; j < m+1; j++){\n",
    "            if (ref[i-1] == read[j-1]){\n",
    "                dp[j][i] = min(dp[j-1][i-1] + match, min(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "            if (ref[i-1] != read[j-1]){\n",
    "                dp[j][i] = min(dp[j-1][i-1] + mismatch, min(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    return dp[m][n];\n",
    "}\n",
    "int main() {\n",
    "    string genome[5];\n",
    "    string dict[] = {\"Reston_genome\", \"Sudan_genome\", \"TaiForest_genome\", \"Zaire_genome\", \"Bundibugyo_genome\"};\n",
    "    string dict2[] = {\"NP\", \"VP35\", \"VP40\", \"GP\", \"VP30\", \"VP24\", \"L\"};\n",
    "    for (int i = 0; i < 5; i++) {\n",
    "        ifstream cin(\"BioProjectFiles/\" + dict[i] + \".fasta\");\n",
    "        string s;\n",
    "        cin >> s;\n",
    "        cin >> s;\n",
    "        while (cin >> s) {\n",
    "            genome[i] += s;\n",
    "        }\n",
    "    }\n",
    "    ifstream cin(\"a.txt\");\n",
    "    string ref, read;\n",
    "    int start, finish;\n",
    "    pair <int,int> indexes[5][7];\n",
    "    for (int i = 0; i < 7; i++){\n",
    "        for(int j = 0; j < 5; j++){\n",
    "        \tcin>>ref>>read>>start>>finish;\n",
    "\t\t\tindexes[j][i] = make_pair(start, finish);\n",
    "        }\n",
    "    }\n",
    "    int edit_distance[7][5][5];\n",
    "    for (int i = 0; i < 7; i++){\n",
    "        ofstream cout(dict2[i] +\".csv\");\n",
    "        for (int j = 0; j < 5; j++){\n",
    "            cout<< dict[j];\n",
    "            if(j != 4){\n",
    "                cout<<\",\";\n",
    "            }\n",
    "        }\n",
    "        cout<<endl;\n",
    "        for(int j = 0; j < 5; j++){\n",
    "            for(int k = 0; k < 5; k++){\n",
    "\t\t\t\tstring ref1 = genome[j].substr(indexes[j][i].first,indexes[j][i].second - indexes[j][i].first + 1);\n",
    "                string ref2 = genome[k].substr(indexes[k][i].first,indexes[k][i].second - indexes[k][i].first + 1);\n",
    "                edit_distance[i][j][k] = aligner(ref1,ref2);\n",
    "           \t\tcout<<edit_distance[i][j][k];\n",
    "                if(k!=4){\n",
    "                    cout<<\",\";\n",
    "                }\n",
    "\t\t\t}\n",
    "\t\t\tcout<<endl;\n",
    "        }\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "نتایج:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GP\n",
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,1069,1101,1088,1063\n",
    "1069,0,1070,1081,1095\n",
    "1101,1070,0,979,825\n",
    "1088,1081,979,0,956\n",
    "1063,1095,825,956,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "L\n",
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,2455,2401,2402,2422\n",
    "2455,0,2404,2404,2434\n",
    "2401,2404,0,2183,1934\n",
    "2402,2404,2183,0,2208\n",
    "2422,2434,1934,2208,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "VP35\n",
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,593,625,601,629\n",
    "593,0,641,600,611\n",
    "625,641,0,547,532\n",
    "601,600,547,0,567\n",
    "629,611,532,567,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "VP40\n",
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,492,706,488,714\n",
    "492,0,710,510,704\n",
    "706,710,0,714,694\n",
    "488,510,714,0,723\n",
    "714,704,694,723,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "VP24\n",
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,469,481,464,486\n",
    "469,0,495,482,493\n",
    "481,495,0,443,422\n",
    "464,482,443,0,432\n",
    "486,493,422,432,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "VP30\n",
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,507,477,460,490\n",
    "507,0,479,481,469\n",
    "477,479,0,422,375\n",
    "460,481,422,0,407\n",
    "490,469,375,407,0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NP\n",
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,955,1002,938,1016\n",
    "955,0,1004,916,1034\n",
    "1002,1004,0,913,821\n",
    "938,916,913,0,933\n",
    "1016,1034,821,933,0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "سپس در بخش ۳.۱ با استفاده از نرم‌افزار R و کتاب‌خانه‌ی phangorn الگوریتم‌های UPGMA و NJ را روی فایل‌های خروجی قسمت قبل پیاده‌سازی کردیم و نتایج را در ادامه آورده‌ایم:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "NP <- read.csv(file = \"~/CLionProjects/untitled/NP.csv\", check.names = FALSE)\n",
    "L <- read.csv(file = \"~/CLionProjects/untitled/L.csv\", check.names = FALSE)\n",
    "GP <- read.csv(file = \"~/CLionProjects/untitled/GP.csv\", check.names = FALSE)\n",
    "VP35 <- read.csv(file = \"~/CLionProjects/untitled/VP35.csv\", check.names = FALSE)\n",
    "VP30 <- read.csv(file = \"~/CLionProjects/untitled/VP30.csv\", check.names = FALSE)\n",
    "VP40 <- read.csv(file = \"~/CLionProjects/untitled/VP40.csv\", check.names = FALSE)\n",
    "VP24 <- read.csv(file = \"~/CLionProjects/untitled/VP24.csv\", check.names = FALSE)\n",
    "Global <- read.csv(file = \"~/CLionProjects/untitled/global.csv\", check.names = FALSE)\n",
    "CSV_3 <- read.csv(file = \"~/CLionProjects/untitled/3_2.csv\", check.names = FALSE)\n",
    "library(phangorn)\n",
    "##L\n",
    "Ltree = upgma(L)\n",
    "neiL = nj(as.dist(L))\n",
    "png(filename=\"~/CLionProjects/untitled/UPGMA_geneL.png\")\n",
    "plot(Ltree)\n",
    "dev.off()\n",
    "png(filename=\"~/CLionProjects/untitled/NJ_geneL.png\")\n",
    "plot(neiL)\n",
    "dev.off()\n",
    "\n",
    "##GP\n",
    "GPtree = upgma(GP)\n",
    "neiGP = nj(as.dist(GP))\n",
    "png(filename=\"~/CLionProjects/untitled/UPGMA_geneGP.png\")\n",
    "plot(GPtree)\n",
    "dev.off()\n",
    "png(filename=\"~/CLionProjects/untitled/NJ_geneGP.png\")\n",
    "plot(neiGP)\n",
    "dev.off()\n",
    "\n",
    "##VP35\n",
    "VP35tree = upgma(VP35)\n",
    "nei35 = nj(as.dist(VP35))\n",
    "png(filename=\"~/CLionProjects/untitled/UPGMA_geneVP35.png\")\n",
    "plot(VP35tree)\n",
    "dev.off()\n",
    "png(filename=\"~/CLionProjects/untitled/NJ_geneVP35.png\")\n",
    "plot(nei35)\n",
    "dev.off()\n",
    "\n",
    "##VP30\n",
    "VP30tree = upgma(VP30)\n",
    "png(filename=\"~/CLionProjects/untitled/UPGMA_geneVP30.png\")\n",
    "plot(VP30tree)\n",
    "dev.off()\n",
    "neiVP30tree = nj(as.dist(VP30))\n",
    "png(filename=\"~/CLionProjects/untitled/NJ_geneVP30.png\")\n",
    "plot(neiVP30tree)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "##VP40\n",
    "VP40tree = upgma(VP40)\n",
    "png(filename=\"~/CLionProjects/untitled/UPGMA_geneVP40.png\")\n",
    "plot(VP40tree)\n",
    "dev.off()\n",
    "neiVP40tree = nj(as.dist(VP40))\n",
    "png(filename=\"~/CLionProjects/untitled/NJ_geneVP40.png\")\n",
    "plot(neiVP40tree)\n",
    "dev.off()\n",
    "\n",
    "##VP24\n",
    "VP24tree = upgma(VP24)\n",
    "png(filename=\"~/CLionProjects/untitled/UPGMA_geneVP24.png\")\n",
    "plot(VP24tree)\n",
    "dev.off()\n",
    "neiVP24tree = nj(as.dist(VP24))\n",
    "png(filename=\"~/CLionProjects/untitled/NJ_geneVP24.png\")\n",
    "plot(neiVP24tree)\n",
    "dev.off()\n",
    "\n",
    "##NP\n",
    "NPtree = upgma(NP)\n",
    "png(filename=\"~/CLionProjects/untitled/UPGMA_geneNP.png\")\n",
    "plot(NPtree)\n",
    "dev.off()\n",
    "neiNPtree = nj(as.dist(NP))\n",
    "png(filename=\"~/CLionProjects/untitled/NJ_geneNP.png\")\n",
    "plot(neiNPtree)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "نتایج:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "UPGMA L\n",
    "<img src=\"UPGMA_geneL.png\">\n",
    "NJ L\n",
    "<img src=\"NJ_geneL.png\">\n",
    "UPGMA NP\n",
    "<img src=\"UPGMA_geneNP.png\">\n",
    "NJ NP\n",
    "<img src=\"NJ_geneNP.png\">\n",
    "UPGMA VP30\n",
    "<img src=\"UPGMA_geneVP30.png\">\n",
    "NJ VP30\n",
    "<img src=\"NJ_geneVP30.png\">\n",
    "UPGMA VP35\n",
    "<img src=\"UPGMA_geneVP35.png\">\n",
    "NJ VP35\n",
    "<img src=\"NJ_geneVP35.png\">\n",
    "UPGMA VP24\n",
    "<img src=\"UPGMA_geneVP24.png\">\n",
    "NJ VP24\n",
    "<img src=\"NJ_geneVP24.png\">\n",
    "UPGMA GP\n",
    "<img src=\"UPGMA_geneGP.png\">\n",
    "NJ GP\n",
    "<img src=\"NJ_geneGP.png\">\n",
    "UPGMA VP40\n",
    "<img src=\"UPGMA_geneVP40.png\">\n",
    "NJ VP40\n",
    "<img src=\"NJ_geneVP40.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "در بخش ۳.۲ به منظور به دست آوردن یک درخت واحد، ابتدا از هر ژنوم مناطقی که ژن‌ها به آن منطبق شده بودند را جدا کرده و به یکدیگر میچسبانیم. نتیجه یک ژنوم جدید خواهد بود که تنها شامل ۷ ناحیه از تطابق ژن هاست سپس این ژنوم‌های جدید را با یکدیگر به صورت سراسری همتراز کرده و فاصله‌ی ویرایش را به دست می‌آوریم سپس درخت‌های مربوطه را با استفاده از‌‌‌r رسم میکنیم.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "UPGMA\n",
    "<img src=\"UPGMA_3-2.png\">\n",
    "NJ\n",
    "<img src=\"NJ_3-2.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "در بخش ۳.۳ با زبان سی یک الگوریتم برای تطابق سراسری بین ژنوم‌های مختلف اجرا میکنیم و سپس نتیجه‌ی فاصله‌های ویرایش را با استفاده از زبان آر به درخت مربوطه تبدیل میکنیم.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#include<iostream>\n",
    "#include<fstream>\n",
    "\n",
    "using namespace std;\n",
    "\n",
    "int dp[19000][19000];\n",
    "\n",
    "int aligner(string ref, string read){\n",
    "    int n = ref.size();\n",
    "    int m = read.size();\n",
    "    int match = 0;\n",
    "    int mismatch = 1;\n",
    "    int gap_open = 1;\n",
    "    for (int i = 0; i < n+1; i++) {\n",
    "        dp[0][i] = i;\n",
    "    }\n",
    "    for (int i = 1; i < m+1; i++) {\n",
    "        dp[i][0] = i;\n",
    "    }\n",
    "    for (int i = 1; i < n+1; i++){\n",
    "        for (int j = 1; j < m+1; j++){\n",
    "            if (ref[i-1] == read[j-1]){\n",
    "                dp[j][i] = min(dp[j-1][i-1] + match, min(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "            if (ref[i-1] != read[j-1]){\n",
    "                dp[j][i] = min(dp[j-1][i-1] + mismatch, min(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    return dp[m][n];\n",
    "}\n",
    "\n",
    "int main() {\n",
    "    string genome[5];\n",
    "    string dict[] = {\"Reston_genome\", \"Sudan_genome\", \"TaiForest_genome\", \"Zaire_genome\", \"Bundibugyo_genome\"};\n",
    "    string dict2[] = {\"NP\", \"VP35\", \"VP40\", \"GP\", \"VP30\", \"VP24\", \"L\"};\n",
    "    for (int i = 0; i < 5; i++) {\n",
    "        ifstream cin(\"BioProjectFiles/\" + dict[i] + \".fasta\");\n",
    "        string s;\n",
    "        cin >> s;\n",
    "        cin >> s;\n",
    "        while (cin >> s) {\n",
    "            genome[i] += s;\n",
    "        }\n",
    "    }\n",
    "    int alignes[5][5];\n",
    "    ofstream cout(\"global.csv\");\n",
    "    for (int j = 0; j < 5; j++){\n",
    "        cout<< dict[j];\n",
    "        if(j != 4){\n",
    "            cout<<\",\";\n",
    "        }\n",
    "    }\n",
    "    cout<<endl;\n",
    "    for (int i = 0; i < 5; i++){\n",
    "        for (int j = i+1; j < 5; j++){\n",
    "            alignes[i][j] = aligner(genome[i], genome[j]);\n",
    "            alignes[j][i] = aligner(genome[i], genome[j]);\n",
    "        }\n",
    "    }\n",
    "    for (int i = 0; i < 5; i++){\n",
    "        for (int j = 0; j < 5; j++){\n",
    "            if(i == j){\n",
    "                cout<<0;\n",
    "            }\n",
    "            else{\n",
    "                cout<<alignes[i][j];\n",
    "            }\n",
    "            if (j!=4){\n",
    "                cout<<\",\";\n",
    "            }\n",
    "        }\n",
    "        cout<<endl;\n",
    "    }\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "خروجی کد بالا:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome\n",
    "0,6788,6789,6697,6797\n",
    "6788,0,6810,6740,6815\n",
    "6789,6810,0,6215,5536\n",
    "6697,6740,6215,0,6247\n",
    "6797,6815,5536,6247,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "UPGMA\n",
    "<img src=\"UPGMA_global.png\">\n",
    "NJ\n",
    "<img src=\"NJ_global.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "در بخش ۳.۴ همانند بخش ۳.۳ الگوریتم همترازی سراسری را اجرا میکنیم، با این تفاوت که این بار ژنوم ماربرگ را هم اضافه کرده ایم.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#include<iostream>\n",
    "#include<fstream>\n",
    "\n",
    "using namespace std;\n",
    "\n",
    "int dp[20000][20000];\n",
    "\n",
    "int aligner(string ref, string read){\n",
    "    int n = ref.size();\n",
    "    int m = read.size();\n",
    "    int match = 0;\n",
    "    int mismatch = 1;\n",
    "    int gap_open = 1;\n",
    "    for (int i = 0; i < n+1; i++) {\n",
    "        dp[0][i] = i;\n",
    "    }\n",
    "    for (int i = 1; i < m+1; i++) {\n",
    "        dp[i][0] = i;\n",
    "    }\n",
    "    for (int i = 1; i < n+1; i++){\n",
    "        for (int j = 1; j < m+1; j++){\n",
    "            if (ref[i-1] == read[j-1]){\n",
    "                dp[j][i] = min(dp[j-1][i-1] + match, min(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "            if (ref[i-1] != read[j-1]){\n",
    "                dp[j][i] = min(dp[j-1][i-1] + mismatch, min(dp[j-1][i] + gap_open, dp[j][i-1] + gap_open));\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    return dp[m][n];\n",
    "}\n",
    "\n",
    "int main() {\n",
    "    string genome[6];\n",
    "    string dict[] = {\"Reston_genome\", \"Sudan_genome\", \"TaiForest_genome\", \"Zaire_genome\",\n",
    "                     \"Bundibugyo_genome\",\"Marburg_genome\"};\n",
    "    string dict2[] = {\"NP\", \"VP35\", \"VP40\", \"GP\", \"VP30\", \"VP24\", \"L\"};\n",
    "    for (int i = 0; i < 5; i++) {\n",
    "        ifstream cin(\"BioProjectFiles/\" + dict[i] + \".fasta\");\n",
    "        string s;\n",
    "        cin >> s;\n",
    "        cin >> s;\n",
    "        while (cin >> s) {\n",
    "            genome[i] += s;\n",
    "        }\n",
    "    }\n",
    "\n",
    "    ifstream cin(\"BioProjectFiles/Marburg_genome.fasta\");\n",
    "    string s;\n",
    "    cin >> s;\n",
    "    cin >> s;\n",
    "    while (cin >> s) {\n",
    "        genome[5] += s;\n",
    "    }\n",
    "    cout<<genome[5].size()<<endl;\n",
    "    int alignes[6][6];\n",
    "    ofstream fout(\"3-4.csv\");\n",
    "    for (int j = 0; j < 6; j++){\n",
    "        fout<< dict[j];\n",
    "        if(j != 5){\n",
    "            fout<<\",\";\n",
    "        }\n",
    "    }\n",
    "    fout<<endl;\n",
    "    for (int i = 0; i < 6; i++){\n",
    "        for (int j = i+1; j < 6; j++){\n",
    "            alignes[i][j] = aligner(genome[i], genome[j]);\n",
    "            alignes[j][i] = alignes[i][j];\n",
    "        }\n",
    "    }\n",
    "    for (int i = 0; i < 6; i++){\n",
    "        for (int j = 0; j < 6; j++){\n",
    "            if(i == j){\n",
    "                fout<<0;\n",
    "            }\n",
    "            else{\n",
    "                fout<<alignes[i][j];\n",
    "            }\n",
    "            if (j!=5){\n",
    "                fout<<\",\";\n",
    "            }\n",
    "        }\n",
    "        fout<<endl;\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "خروجی کد بالا:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reston_genome,Sudan_genome,TaiForest_genome,Zaire_genome,Bundibugyo_genome,Marburg_genome\n",
    "0,6788,6789,6697,6797,8926\n",
    "6788,0,6810,6740,6815,8989\n",
    "6789,6810,0,6215,5536,9052\n",
    "6697,6740,6215,0,6247,8986\n",
    "6797,6815,5536,6247,0,9064\n",
    "8926,8989,9052,8986,9064,0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "سپس با استفاده از نرم افزار آر درخت مربوط به جدول فاصله‌ی بالا را میکشیم. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
